src/FsAutoComplete.Core/AdaptiveExtensions.fs

```diff
@@ -6,40 +6,47 @@ open FSharp.Data.Traceable
 open System.Threading.Tasks
 open IcedTasks
 open System.Threading
-
+open FsAutoComplete
+open FsAutoComplete.Logging
+open FsAutoComplete.Logging.Types
 
 [<AutoOpen>]
 module AdaptiveExtensions =
+  let rec logger = LogProvider.getLoggerByQuotation <@ logger @>
+  open System.Runtime.ExceptionServices
 
   type CancellationTokenSource with
 
-    /// Communicates a request for cancellation. Ignores ObjectDisposedException
     member cts.TryCancel() =
       try
-        cts.Cancel()
-      with :? ObjectDisposedException ->
-        ()
+        if not <| isNull cts then
+          cts.Cancel()
+      with
+      | :? ObjectDisposedException
+      | :? NullReferenceException -> ()
 
-    /// Releases all resources used by the current instance of the System.Threading.CancellationTokenSource class.
     member cts.TryDispose() =
-      try
+      // try
+      if not <| isNull cts then
         cts.Dispose()
-      with _ ->
-        ()
+  // with _ -> ()
+
 
   type TaskCompletionSource<'a> with
 
     /// https://github.com/dotnet/runtime/issues/47998
     member tcs.TrySetFromTask(real: Task<'a>) =
-      task {
-        try
-          let! r = real
-          tcs.TrySetResult r |> ignore<bool>
-        with
-        | :? OperationCanceledException as x -> tcs.TrySetCanceled(x.CancellationToken) |> ignore<bool>
-        | ex -> tcs.TrySetException ex |> ignore<bool>
-      }
-      |> ignore<Task<unit>>
+
+      real.ContinueWith(fun (task: Task<_>) ->
+        match task.Status with
+        | TaskStatus.RanToCompletion -> tcs.TrySetResult task.Result |> ignore<bool>
+        | TaskStatus.Canceled ->
+          tcs.TrySetCanceled(TaskCanceledException(task).CancellationToken)
+          |> ignore<bool>
+        | TaskStatus.Faulted -> tcs.TrySetException(task.Exception.InnerExceptions) |> ignore<bool>
+
+        | _ -> ())
+      |> ignore<Task>
```

Comment:
Changed back to `ContinueWith` style as getting stack traces weren't working properly. Wish I had a good idea how to make a test for that.